File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        def dfs(choices, path):
            if len(choices) == 0:
                result.append(path.copy())
                return
            
            choice = choices[0]
            remChoices = choices[1:]

            dfs(remChoices, path)
            path.append(choice)
            dfs(remChoices, path)
            path.pop()
        
        result = []
        dfs(nums, [])

        return result

File: 0078-subsets/0078-subsets.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var subsets = function (nums) {
    const dfs = function (choices, path) {
        if (choices.length === 0) {
            result.push([...path]);
            return;
        }

        const choice = choices[0];
        const remainingChoices = choices.slice(1);

        dfs(remainingChoices, path);
        path.push(choice);
        dfs(remainingChoices, path);
        path.pop();
    }

    const result = [];
    dfs(nums, []);

    return result;
};